autoPyTorch/evaluation/abstract_evaluator.py

```diff
@@ -0,0 +1,533 @@
+        if self.budget_type is not None:
+            if self.budget_type == 'runtime':
+                default_pipeline_options['runtime'] = self.budget
+                if 'epochs' in default_pipeline_options:
+                    del default_pipeline_options['epochs']
+            elif self.budget_type == 'epochs':
+                default_pipeline_options['epochs'] = self.budget
+                if 'runtime' in default_pipeline_options:
+                    del default_pipeline_options['runtime']
+            default_pipeline_options['budget_type'] = self.budget_type
+        self.logger.debug("Default pipeline options: {}".format(default_pipeline_options))
+        self._init_params.update({**default_pipeline_options})
+
+    def _get_model(self) -> BaseEstimator:
+        assert self.model_class is not None, "Can't return model, model_class not initialised"
+        if not isinstance(self.configuration, Configuration):
+            model = self.model_class(config=self.configuration,
+                                     random_state=np.random.RandomState(self.seed),
+                                     init_params=self._init_params)
+        else:
+            if self.configuration is not None:
+                self.logger.debug("model_class is : {}".format(self.model_class))
+            model = self.model_class(config=self.configuration,
+                                     dataset_properties=self.dataset_properties,
+                                     random_state=np.random.RandomState(self.seed),
+                                     include=self.include,
+                                     exclude=self.exclude,
+                                     init_params=self._init_params)
+        return model
+
+    def _loss(self, y_true: np.ndarray, y_hat: np.ndarray) -> Dict[str, float]:
+        """Auto-sklearn follows a minimization goal, so the make_scorer
+        sign is used as a guide to obtain the value to reduce.
+
+        On this regard, to optimize a metric:
+            1- score is calculared with calculate_score, with the caveat, that if
+            for the metric greater is not better, a negative score is returned.
+            2- the err (the optimization goal) is then:
+                optimum - (metric.sign * actual_score)
+                For accuracy for example: optimum(1) - (+1 * actual score)
+                For logloss for example: optimum(0) - (-1 * actual score)
+        """
+
+        if not isinstance(self.configuration, Configuration):
+            return {self.metric[0].name: 1.0}
+
+        score = calculate_score(
+            y_true, y_hat, self.task_type, self.metric)
+
+        err = {metric.name: metric._optimum - score[metric.name] for metric in self.metric
+               if metric.name in score.keys()}
+
+        return err
+
+    def finish_up(self, loss: Dict[str, float], train_loss: Dict[str, float],
+                  opt_pred: np.ndarray, valid_pred: np.ndarray,
```

Comment:
Is `opt_pred` simply `train_pred`?

Reply:
opt_pred is the predictions on the val split from the resampling strategy, and valid_pred is predictions on the validation set if provided by the user, same for test_pred